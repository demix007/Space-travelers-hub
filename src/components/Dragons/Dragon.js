import React from 'react';
import '../../styles/dragons.css';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { reserveDragon, cancelReservation } from '../../redux/dragons';

const Dragon = (props) => {
  const dispatch = useDispatch();
  const {
    dragonName, type, dragonImg, dragonId, reserved,
  } = props;
  return (
    <div className="dragon-card">

      <img src={dragonImg} className="dragon-img" alt="dragon pic" />

      <div className="dragon-desc">
        <h3 className="dragon-name">
          {dragonName}
        </h3>
        <p>
          {reserved && (
          <span className="reserve-badge">Reserved</span>
          )}
          {type}
        </p>
        {
          reserved ? (
            <button
              type="button"
              className="cancel-btn"
              onClick={() => {
                dispatch(cancelReservation(dragonId));
              }}
            >
              Cancel Reservation
            </button>
          ) : (
            <button
              type="button"
              className="reserve-btn"
              onClick={() => {
                dispatch(reserveDragon(dragonId));
              }}
            >
              Reserve Dragon
            </button>
          )
        }
      </div>
    </div>
  );
};
Dragon.propTypes = {
  dragonName: PropTypes.string,
  type: PropTypes.string,
  dragonImg: PropTypes.string,
  dragonId: PropTypes.string,
  reserved: PropTypes.bool,
}.isRequired;
export default Dragon;
