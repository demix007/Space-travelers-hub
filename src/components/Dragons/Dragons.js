import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Dragon from './Dragon';
import '../../styles/dragons.css';
import { fetchDragons } from '../../redux/dragons';

const Dragons = () => {
  const dispatch = useDispatch();
  const shouldFetch = useRef(true);
  const dragonsArr = useSelector((state) => state.dragons);

  useEffect(() => {
    if (shouldFetch.current) {
      if (dragonsArr.length === 0) {
        shouldFetch.current = false;
        dispatch(fetchDragons());
      }
    }
  }, []);
  if (dragonsArr.length > 0) {
    return (
      <div className="dragons-container">
        {
          dragonsArr.map((dragon) => (
            <Dragon
              key={dragon.id}
              dragonName={dragon.dragon_name}
              type={dragon.type}
              dragonImg={dragon.flickr_images}
              dragonId={dragon.id}
              reserved={dragon.reserved}
            />
          ))
        }
      </div>
    );
  }
  return '';
};
export default Dragons;
