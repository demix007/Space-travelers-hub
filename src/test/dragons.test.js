import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../redux/configureStore';
import Dragon from '../components/Dragons/Dragon';
import Dragons from '../components/Dragons/Dragons';

describe('Dragons', () => {
  it('Dragon component is rendered correctly', () => {
    const dragon = render(
      <Provider store={store}>
        <Dragon />
      </Provider>,
    );
    expect(dragon).toMatchSnapshot();
  });

  it('Dragons component rendered correctly', () => {
    const dragons = render(
      <Provider store={store}>
        <Dragons />
      </Provider>,
    );
    expect(dragons).toMatchSnapshot();
  });
});
