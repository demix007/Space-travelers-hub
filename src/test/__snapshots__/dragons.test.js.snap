// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dragons Dragon component is rendered correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="dragon-card"
      >
        <img
          alt="dragon pic"
          class="dragon-img"
        />
        <div
          class="dragon-desc"
        >
          <h3
            class="dragon-name"
          />
          <p />
          <button
            class="reserve-btn"
            type="button"
          >
            Reserve Dragon
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="dragon-card"
    >
      <img
        alt="dragon pic"
        class="dragon-img"
      />
      <div
        class="dragon-desc"
      >
        <h3
          class="dragon-name"
        />
        <p />
        <button
          class="reserve-btn"
          type="button"
        >
          Reserve Dragon
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Dragons Dragons component rendered correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
